community, wg, template: create wg templates

dhiller opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:

This PR:
* updates the document GOVERNANCE.md that describes SIGs and WGs.
* creates a template for working group definitions, aligned to how SIG templates currently look.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

/cc @oshoval @fabiand

**Checklist**

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note

```

dhiller commented:
/cc @iholder101 @lyarwood @chandramerla

fabiand commented:
Thanks, Daniel!

dhiller commented:
/cc @aburdenthehand

lyarwood commented on SIGS_AND_WGS.md:
```diff
@@ -0,0 +1,28 @@
+
+## WG
+
+A **WG (Working Group)** is a group of people that owns a more specific topic within the project, possibly touching the scope of more than one SIG, with whom the WG interacts.
```
I'm against redefining the k8s meaning of a WG here by suggesting that ownership is possible outside of a SIG [1][2].

As I've suggested elsewhere [3] we can avoid this by just using subprojects within a SIG [4] and keep the k8s definition of a WG to be a multi SIG collaboration vehicle. 

[1] https://github.com/kubernetes/community/blob/master/governance.md#working-groups
[2] https://github.com/kubernetes/community/blob/master/committee-steering/governance/wg-governance.md#is-it-a-working-group-yes-if
[3] https://github.com/kubevirt/community/pull/299#issuecomment-2149750036
[4] https://github.com/kubernetes/community/blob/master/governance.md#subprojects

dhiller replied:
@lyarwood in Kubernetes WGs are short-lived, and I don't think that this serves node architecture related matters well (for which the initial PR was created), since for that matter clearly an ownership needs to exist in the longer term.

Therefore I have refrained from copying the k8s definitions.

Aside of that I am not talking about specific code ownership here (I agree that this is restricted to SIGs IMHO), but rather about topic ownership. Maybe since I am not a native speaker this is just a matter of words, so I could use some advice about how to rephrase it.

lyarwood replied:
@dhiller processes and deliverables are all also ultimately owned by a SIG within k8s thus my suggestion again to use a SIG sub project instead of redefining the k8s WG definition. These sub projects also have leads [1] and owners [2] so should accommodate what you want to achieve in  https://github.com/kubevirt/community/pull/298. I'm working through an example for instance types under SIG Compute within https://github.com/kubevirt/kubevirt/pull/12062 and https://github.com/kubevirt/community/pull/302 FWIW.

dhiller replied:
First of all: what is the problem with redefining WGs for KubeVirt? 

Second: Actually, when looking at the document you provided, the sense of the above sentence isn't even redefining it, but rather just avoiding the phrase "short-lived", no?

In that way, and also looking at the document, I think that working group is perfectly right for node architecture, since it actually spans multiple SIGs.

And the sentence explicitly says "owns a more specific topic" and says nothing about code ownership, so again - what is wrong with that sentence?

Please help, thanks :)

lyarwood replied:
> First of all: what is the problem with redefining WGs for KubeVirt?

I just don't understand why it's required when the existing and well understood (at least across CNCF) definitions of k8s provide everything we need IMHO.

> Second: Actually, when looking at the document you provided, the sense of the above sentence isn't even redefining it, but rather just avoiding the phrase "short-lived", no?

That and suggesting ownership of a topic instead of being targeted at resolving a specific problem. 

> In that way, and also looking at the document, I think that working group is perfectly right for node architecture, since it actually spans multiple SIGs.

They definitely are for organising around a defined problem like initially standing up a new architecture across multiple SIGs yes but the on-going ownership of the topic, processes and deliverables should still reside in a SIG IMHO with specific architectures listed as sub projects.

> And the sentence explicitly says "owns a more specific topic" and says nothing about code ownership, so again - what is wrong with that sentence?

As I tried to say before SIGs own more than just code and WGs in k8s own nothing.

> Please help, thanks :)

FWIW I've not received any positive feedback on my stance here or in other PRs so it's likely that I'm standing alone on this hill.

lyarwood commented on wg-TEMPLATE/wg-charter-template.md:
```diff
@@ -0,0 +1,31 @@
+
+### In scope
+
+#### Code, Binaries and Services
```
Can you drop this?

lyarwood reviewed (commented):
Looks good, I'd still like to drop the `Code, Binaries and Services` section and just leave the generic in and out of scope parts.

EdDev reviewed (commented):
IIUC the original reasoning to introduce the WG concept originated from the need to formalize code ownership of smaller scoped groups (than SIGs).

After @lyarwood clarified the difference between a WG and a subproject, the original reasoning is lost.

IMO there is a need to clarify the new reasoning to establish WG in Kubevirt with examples to such groups that are needed.
At the moment, I’m not familiar with a group of members that meet and discuss cross-SIG topics (that is not owned by a specific SIG). I there something like this?

On the other hand, the original need for a more granular code ownership is still needed. In that regard, I agree with @lyarwood that we need to define subprojects in Kubevirt.

lyarwood commented:
> IIUC the original reasoning to introduce the WG concept originated from the need to formalize code ownership of smaller scoped groups (than SIGs).
> 
> After @lyarwood clarified the difference between a WG and a subproject, the original reasoning is lost.
> 
> IMO there is a need to clarify the new reasoning to establish WG in Kubevirt with examples to such groups that are needed. At the moment, I’m not familiar with a group of members that meet and discuss cross-SIG topics (that is not owned by a specific SIG). Is there something like this?

I posted a suggestion on the ML about this yesterday using SIG sub projects for ownership and WGs to drive standing up each architecture across other SIGs:

https://groups.google.com/g/kubevirt-dev/c/G6eCHpxJ9DU/m/PaeGjNSaAAAJ

> - A subproject per arch under SIG buildsystem to own the build aspect
but also the infra associated with building, testing etc.
> - A working group per arch ran by the SIG buildsystem subproject team
with the aim of handling the cross SIG collaboration required to stand
up support for said arch
> - Additional subprojects per arch in any SIG where it's required to
own specific logic, I could see this being useful in SIG compute for
example to handle arch specific logic

I might be overthinking things at this point but if we do need WGs going forward this pattern could work.

> On the other hand, the original need for a more granular code ownership is still needed. In that regard, I agree with @lyarwood that we need to define subprojects in Kubevirt.

Thanks :)

EdDev commented:
> WGs to drive standing up each architecture across other SIGs

I see, thank you for the ref.
It will be useful to evaluate this after experiencing the actual work efforts.

dhiller commented:
@EdDev @lyarwood I see this PR as a ground work on which the actual folks from which we require responsibilities wrt code and infra can build upon. Therefore I've kept this pretty vague by intention.

@lyarwood I don't want to take the decision from the folks inside the arch teams for S390X or ARM how they organize themselves- I think that is a decision to be taken by them; in claiming the responsibility by joining a WG and assigning subsets of code to subproject `sig-buildsystem-arch-*`.

What I would want to hear from folks @jschintag @cfilleke @vamsikrishna-siddu @zhlhahaha and others is how they are going to organize themselves in keeping the responsibilities that can be seen around arch work. We might then adjust the follow up PR as needed - this one I think might be left as is currently...

WDYT?

lyarwood reviewed (approved):
> @EdDev @lyarwood I see this PR as a ground work on which the actual folks from which we require responsibilities wrt code and infra can build upon. Therefore I've kept this pretty vague by intention.
> 
> @lyarwood I don't want to take the decision from the folks inside the arch teams for S390X or ARM how they organize themselves- I think that is a decision to be taken by them; in claiming the responsibility by joining a WG and assigning subsets of code to subproject `sig-buildsystem-arch-*`.

ACK understood.

/lgtm

> What I would want to hear from folks @jschintag @cfilleke @vamsikrishna-siddu @zhlhahaha and others is how they are going to organize themselves in keeping the responsibilities that can be seen around arch work. We might then adjust the follow up PR as needed - this one I think might be left as is currently...
> 
> WDYT?

+1

vamsikrishna-siddu commented:
Thanks @dhiller 
/lgtm

chandramerla commented on SIGS_AND_WGS.md:
```diff
@@ -0,0 +1,25 @@
+Examples:
+* [sig-network - charter](./sig-network/charter.md)
+* [sig-network - kubevirt/kubevirt OWNERS_ALIASES](https://github.com/kubevirt/kubevirt/blob/a7e0311d8704663351abd4bc9bbc8511753d2838/OWNERS_ALIASES#L60)
+* [sig-ci - sigs.yaml](https://github.com/kubevirt/community/blob/4f63a79c0ed810aa332cd6716d4986001d28bcd7/sigs.yaml#L119)
```
This link seems to be broken.

![image](https://github.com/kubevirt/community/assets/156769580/225d8907-bc51-4194-ac2c-715b1be7e495)

chandramerla commented on SIGS_AND_WGS.md:
```diff
@@ -0,0 +1,25 @@
+
+## [SIGs] (Special Interest Groups)
+
+Our [SIGs] define mission and scope via their charter, ownership of code via [OWNERS](https://www.kubernetes.dev/docs/guide/owners/), and define members and roles in community [sigs.yaml].
```
sigs.yaml hyperlink in above sentence seem to be broken

![image](https://github.com/kubevirt/community/assets/156769580/225d8907-bc51-4194-ac2c-715b1be7e495)

dhiller replied:
IIUC the link doesn't work in the preview since it's not merged yet - or am I mistaken?

chandramerla replied:
@dhiller The hyperlink sigs.yaml is broken in line #7 and in line #25, but where as in line #12, sigs.yaml hyperlink is working. So it seemed like the hyperlink in those two places is broken (probably because there isn't any URL attached to hyperlink? )

dhiller replied:
Since I am using reference links, the reference for the link is defined in the bottom section in order to avoid repetitions. Please see [the markdown spec](https://daringfireball.net/projects/markdown/syntax#link) 

In line 12 I am referring to a specific line, therefore I avoided using the reference only.

chandramerla replied:
@dhiller Thank you for patiently explaining me about this and sharing the link. Now I understand the markdown better :)
By the way this file when opened from your [branch](https://github.com/dhiller/community/blob/wg-template/SIGS_AND_WGS.md) the links are working fine

dhiller replied:
@chandramerla you're welcome :slightly_smiling_face: 

Thanks for your comment about my branch - good point, I should have included that in my comment.

chandramerla reviewed (commented):
/lgtm

jschintag commented:
Thank you @dhiller.

From s390x side
/lgtm

zhlhahaha commented:
/lgtm
From Arm side, thanks @dhiller

aburdenthehand commented on SIGS_AND_WGS.md:
```diff
@@ -0,0 +1,25 @@
+# SIGs and WGs
```
I'm ambivalent about separate files, but if we do separate I see no need to separate sub-projects from SIGs. 
I agree with your subsequent points. @dhiller, I can help summarise some key paras from the k8s docs if you'd like

dhiller replied:
I agree that usage of subprojects is missing here, I'll update the PR wrt that. I am not sure whether subprojects should be inside the filename though. I'd be against splitting into separate files, since having sections does a good enough job - in contrast to having a file WG.md where only one sentence is contained. WDYT?

To the latter point - WDYM? Isn't referencing the Kubernetes model as seen below enough? Why should we repeat the definitions?

dhiller replied:
^^ @EdDev

dhiller replied:
^^ @EdDev

aburdenthehand replied:
I can't speak for Ed but imho something along the lines of :
"We follow the definition of a SIG: A SIG is comprised of members from multiple companies and organizations, with a common purpose of advancing the project with respect to a specific topic, such as Networking or Documentation. Our goal is to enable a distributed decision structure and code ownership, as well as providing focused forums for getting work done, making decisions, and onboarding new contributors.
For more information, see the k8s docs"
Or similar.

The reason being is that it gives the casual reader the context for what this is all about without having to follow links, which means they can understand the page on its own merits, and the links are there for them to follow fro greater/more specific understanding.
I don't expect the k8s descriptions are likely to change all that often so I don't really see an issue with maintenance.

EdDev replied:
>  am not sure whether subprojects should be inside the filename though. I'd be against splitting into separate files, since having sections does a good enough job - in contrast to having a file WG.md where only one sentence is contained. WDYT?

SIGs and WG are IMO separated topics and better appear in different files.
I agree that subprojects fit the SIG file as they are part of it.
The current name is very odd IMO, trying to name two different things in one name.

dhiller replied:
^^ @aburdenthehand @EdDev WDYT?

@aburdenthehand what you said makes sense - a short collection of what's this about makes sense

aburdenthehand replied:
@dhiller We currently have a Governance.md file in the repo: it is largely, but not exclusively!, concerned with governance of project maintainers. Yeah, maybe that would be a better place for this content (summary of SIGs and WGs). 
We can also update the Readme link to point to their particular sections in the governance doc.

dhiller commented:
@aburdenthehand @EdDev what you said makes sense - I've restructured the `GOVERNANCE.md` to include the paragraphs about SIGS, subprojects and WGs, also I've added a subset of the descriptions to each paragraph, and added links to k8s resources about the topics.

PTAL, thank you all for your feedback!

lyarwood commented on GOVERNANCE.md:
```diff
@@ -105,3 +108,60 @@ request a vote be taken.
+may have a single subproject, but many SIGs have multiple significant subprojects
+with distinct (though sometimes overlapping) sets of contributors and owners, who
+act as subproject’s technical leaders: responsible for vision and direction and
+overall design, choose/approve change proposal (KEP) approvers, field technical
```
We don't use `KEPs` just yet so replace with design proposals I guess?

lyarwood commented on GOVERNANCE.md:
```diff
@@ -105,3 +108,60 @@ request a vote be taken.
+that are short-lived or that span multiple SIGs.
+
+Working groups are primarily used to facilitate topics of discussion that are in
+scope for Kubernetes but that cross SIG lines.
```
KubeVirt

lyarwood reviewed (changes requested):
Thanks @dhiller this looks great, just a few simple nits.

dhiller commented:
@lyarwood thanks for the review, updated, PTAL!

lyarwood commented:
/lgtm

aburdenthehand commented on GOVERNANCE.md:
```diff
@@ -105,3 +108,60 @@ request a vote be taken.
+[Kubernetes SIGs]: https://github.com/kubernetes/community/blob/master/governance.md#sigs
+[Kubernetes Subprojects]: https://github.com/kubernetes/community/blob/master/governance.md#subprojects
+[Kubernetes WGs]: https://github.com/kubernetes/community/blob/master/governance.md#working-groups
+[OARP]: https://stumblingabout.com/tag/oarp/
```
I like the slide this links to I'm not sure it's a good source to link to: a short-lived blog that hasn't been updated in 11 years. Unfortunately, I also can't find an alternative for OARP, and I wonder if it's because it's a somewhat bespoke modification of RACI etc since the blog owner was a strategy consultant. 
The linked blogpost doesn't really go into much detail either beyond the slide. 
What about linking to our Community membership doc? https://github.com/kubevirt/community/blob/main/membership_policy.md
In the future we could add to that doc with a diagram showing the relationship between the different roles.  WDYT?

dhiller replied:
Yeah, good point. I agree that the article might vanish or even be modified at some later point in the future, so I am going to link to the membership policy for now, knowing that the roles are still missing - I'll create a followup PR that updates the membership policy with the additional roles.

@aburdenthehand would that work for you?

aburdenthehand replied:
@dhiller Yeah, I think that will work better. Thanks Daniel!

aburdenthehand reviewed (commented):
Great stuff @dhiller :D 
Thanks for putting all this together. I have a question to consider and a teensy tiny nit.

aburdenthehand commented on wg-TEMPLATE/README.md:
```diff
@@ -0,0 +1,37 @@
+
+All KubeVirt WGs must define a charter.
+
+- The scope must define
+  - what problem the WG is solving
+  - what deliverable(s) will be produced to whom
+  - when the problem is solved
+- The governance must outline
+  - the responsibilities within the WG as well as
+  - the roles owning those responsibilities.
+  - which SIGs are stakeholders
```
```suggestion
- The scope must define:
  - what problem the WG is solving
  - what deliverable(s) will be produced, and to whom
  - when the problem is expected to be resolved
- The governance must outline:
  - the responsibilities within the WG
  - the roles owning those responsibilities
  - which SIGs are stakeholders
```

aburdenthehand commented on wg-TEMPLATE/wg-charter-template.md:
```diff
@@ -0,0 +1,24 @@
+### Stakeholder SIGs
+
+A list of SIGs that are relevant to the WG.
+The template may also mention
+* which SIGs are more relevant than others,
+* which subprojects are related to the WG
```
```suggestion
The template may also mention:
* which SIGs are more relevant than others, and
* which subprojects are related to the WG
```

aburdenthehand commented on wg-TEMPLATE/wg-charter-template.md:
```diff
@@ -0,0 +1,24 @@
+### Roles and responsibilities
+
+#### Chairs
+- @your-github-handle
```
Did we want two names for Chair and Leads?

dhiller replied:
In K8s terms sigs and wgs have chairs, while subprojects have leads - or did I misread that?

dhiller replied:
> Did we want two names for Chair and Leads?

Ah, now I get it - I had suggested that we want two, since otherwise there'd be an interval of vacancy during PTO.

dhiller replied:
Actually I am not really sure whether we even need the lead role inside this template.

aburdenthehand reviewed (commented):
Great stuff. I have no complaints and am really happy to see this.
I've suggested a couple of changes in the templates

dhiller commented:
> Great stuff. I have no complaints and am really happy to see this. I've suggested a couple of changes in the templates

@aburdenthehand thank you for your review, I've addressed the comments. PTAL!

aburdenthehand commented:
/lgtm

dhiller commented:
This PR is waiting for three weeks now.

Hey @cwilkers @davidvossel @fabiand @rmohr @vladikr :wave: 

Since @aburdenthehand is out for a while, is one of you able to take a quick look and approve this?

vladikr commented:
@jean-edouard did you have a chance to go over this PR?

jean-edouard commented:
> @jean-edouard did you have a chance to go over this PR?

@dhiller are WGs still something we're considering?

dhiller commented:
> > @jean-edouard did you have a chance to go over this PR?
> 
> @dhiller are WGs still something we're considering?

@jean-edouard this PR only does groundwork, i.e. laying out the structure for contributors who want to organize inside a working group, in cases where a SIG does not fit.

IMHO this is necessary so that contributors know how we think governance inside KubeVirt org works - and I believe that a defined governance process is also a requirement for the graduation of KubeVirt, see https://github.com/kubevirt/community/issues/307.

Please note that this PR does not define any WG by itself - although I did create [a PR for WG arches](https://github.com/kubevirt/community/pull/298) to test out how an example WG might look like.

EdDev reviewed (approved):
Looks great, thank you very much!

jean-edouard commented on wg-TEMPLATE/README.md:
```diff
@@ -0,0 +1,37 @@
+
+## Goals
+
+The primary goal of the charters is to define the scope of the WG within KubeVirt and how the WG can determine whether it has achieved their solution. A majority of the effort should be spent on these concerns.
```
```suggestion
The primary goal of the charters is to define the scope of the WG within KubeVirt and how the WG can determine whether it has reached its goal(s). Most of the effort should be spent on these concerns.
```

vladikr commented:
Thanks @dhiller !
/approve

jean-edouard reviewed (commented):
/approve
Just one nit inline.
